# DS 3000 HW 1

Due: Tues Feb 02 @ 11:59 PM EST

### Submission Instructions
Submit this `ipynb` file to gradescope.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run "Kernel > Restart & Run All" just before uploading the `ipynb` file to gradescope.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or post a piazza note if you have a specific question about your work you'd like another pair of eyes to talk through.  (Remember, mark your piazza note private if it contains anything which may be considered a solution to the exercise).

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment.  For example, a python trick I find particularly clever (and useful, sometimes):

```python
from collections import defaultdict

def tree(): 
    # https://gist.github.com/hrldcpr/2012250
    return defaultdict(tree)
```

You need not cite the official python documentation or the documentation of any python library which is imported in the template (e.g. matplotlib, numpy, scipy).

**Documentation / style counts for credit**  Please see our course's python style guide, available on canvas, for further information.

# Markdown

## Part 1 (15 points: (2, 2, 3, 2, 2, 4) pts per item in list below):
Use the markdown language below to create your own brief wikipedia-esque description of any subject of interest. 

Your mini-wiki page must include:
- three headers: a title, subtitle and subsubtitle (the #, ##, ### syntax)
- an embedded image from a web address (use an [image hosting site](https://makeawebsitehub.com/free-photo-hosting/) if you'd like to upload your own)
- a table of size at least 3 rows x 3 columns (needn't be correct, but must make sense)
- a list
- a link to another website
- a link to another jupyter cell

Please be brief in your text.  Aim for roughly 3 sentences total of text.  We won't grade based on content, but keep it class appropriate.  If you make the grader smile, no extra credit will be awarded beyond the satisfaction of having made somebody's day better :)

In [1]:
# please add cells as needed to put your answer directly below each corresponding Part

# the introduction of stock
## Stock derivatives
### stock index

## Images
![alt text](https://www.cmegroup.com/education/images/articles/2018/stock-index-spread-opportunities-fig01.png)



| name of stock |      |    |      |
|---------------|------|----|------|
| skyHighcorp   | 3243 | 76 | 21.2 |
| lowDownInc    | 2735 | 18 | 41   |
| ValueNowInc   | 1894 | 12 | 22   |



## Lists
here is a list of top stock:
- Apple Inc
- Tesla Inc
- FaceBook Inc


## Links
You can link to a nasdaq website from  [this one](https://www.nasdaq.com/). 


You can link to [another cell](#another_cell)

# this is another topic about the stock basic knowledge
<a id='another_cell'></a>

# Prisoner's Dilemma
This section asks you to implement a game of repeated [prisoner's dilemma](https://en.wikipedia.org/wiki/Prisoner%27s_dilemma).   

Prisoner's dilemma is a cooperation game.  Two players simultaneously tell if they will cooperate or not in a given round.  Points are awarded based on both player's cooperation (or not) decisions:
- If both players cooperate with each other, each receives 2 points
- If both players do not cooperate with each other, each receives 1 point
- If one player cooperates while the other does not, the cooperating player gets 0 points, the non-cooperator receives 3 points

Repeated prisoner's dilemma has the two players continue to play rounds until one player reaches some threshold of points (this threshold is stored as `score_to_win` below).  

What makes repeated prisoner's dilemma so fun is that the history of your opponents decisions allow you to make an informed decision about whether you should cooperate or not in the current round.  In this implementation, we'll allow each player to choose to cooperate or not in a new round based exclusively on their opponents choice in the last round.

## Part 2 (10 points): Copycat strategy
Implement the `get_coop_copycat()` function below which copies the opponents last strategy.  Be sure to write the docstring yourself (for a simple function like this one, documentation is a worth a large portion of the credit ... please use the style guide!).

In [2]:
def get_coop_copycat(opp_last_coop=True):
    """
    Args:
    opp_last_coop: is a boolean that whether the previous 
    player choose to cooperate or not. 

    returns:
    opp_last_coop: return the previous one's choice as a boolean.
    """

    return opp_last_coop



In [3]:
# test cases
assert get_coop_copycat(opp_last_coop=True) == True
assert get_coop_copycat(opp_last_coop=False) == False

## Part 3 (13 points): Friendly strategy
Implement the `get_coop_friendly()` function below which ignores the opponents last strategy and chooses to cooperate 90% of the time.  (Use [np.random.choice()](https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html) to select a random number).  Be sure to write the docstring yourself.

In [4]:
import numpy as np

In [5]:
def get_coop_friendly():
    """
    return:
    a random number from 0, 1, where 0 represent that 
    status that the player chooose not to be friendly coop, while 1
    represent a coop-friendly. 
    """
    return np.random.choice(2,1,p=[0.1,0.9])

   

## Part 4 (6 points): Testing Friendly Strategy
One way to test if our `get_coop_friendly` function works is to ensure that if we run it many times the result is close to its expected value.  In other words, our friendly player should produce about 9k (i.e. 9000) cooperates given 10k rounds of playing.  We'll say the function works if it outputs between 8k and 10k cooperations.

Write a loop which counts the number of cooperations `get_coop_friendly()` produces in 10k rounds.  Use `assert` to check that this number of cooperations is indeed in the range we expect.

In [6]:

#this is the script to check while we run the game for 10000 times, what will the friendly strategy output
# the reuslt should in the range of 8000 to 10000

#record is used to record the total times while the function return 1, which is friendly_coop
record = 0
for i in range(10000):
    record = record + get_coop_friendly()[0]

    


#test whether the record is in the range of 8000 to 10000
assert record in range(8000, 10000)




## Part 5 (21 points): Scoring
Complete the `score_pdil()` function below to assign scores based on the scoring list directly above.

Use the docstring and test cases to inform your implementation.  If `verbose=True`, you should print one of the following messages:
- both players cooperate
- both players do not cooperate
- one player cooperates, the other does not

In [7]:
def score_pdil(p0_coop, p1_coop, verbose=False):
    """ scores a single round of prisoner's dilemma
    
    Args:
        p0_coop (bool): True if player 0 cooperates,
            otherwise false
        p1_coop (bool): True if player 1 cooperates,
            otherwise false
        verbose (bool): toggles output to command line
            
    Returns:
        p0_points (int): "points" earned by p0 this round
        p1_points (int): "points" earned by p1 this round
    """
    # cast both inputs to boolean
    p0_coop = bool(p0_coop)
    p1_coop = bool(p1_coop)

    # the three potential answers
    answers = ["both players cooperate","both players do not cooperate","one player cooperates, the other does not"]

    # initialize the points of each to zero
    p0_points = 0
    p1_points = 0

    # if the two both cooperate, then both get 2 poists
    if p0_coop == True and p1_coop == True:
        p0_points = 2
        p1_points = 2
        if verbose == True:
            return answers[0]

    # if only one coperate, then the one coopearte get zero while the other one get 3 points
    elif p0_coop == True and p1_coop == False:
        p0_points = 0
        p1_points = 3
        if verbose == True:
            return answers[2]
        

    # if only one coperate, then the one coopearte get zero while the other one get 3 points
    elif p0_coop == False and p1_coop == True:
        p0_points = 3
        p1_points = 0
        if verbose == True: 
            return answers[2]
       
    # if neither of them choose to cooperate, then each of them get 1 points.
    elif p0_coop == False and p1_coop == False:
        p0_points = 1
        p1_points = 1
        if verbose == True:
            return answers[1]

    return p0_points, p1_points
        

    
    

In [8]:
# test cases
assert score_pdil(p0_coop=True, p1_coop=True) == (2, 2)
assert score_pdil(p0_coop=True, p1_coop=False) == (0, 3)
assert score_pdil(p0_coop=False, p1_coop=True) == (3, 0)
assert score_pdil(p0_coop=False, p1_coop=False) == (1, 1)

# my own test here, which is to check while the verbose is true:
assert score_pdil(p0_coop=True, p1_coop= True, verbose= True) == "both players cooperate"
assert score_pdil(p0_coop=True, p1_coop= False, verbose= True) == "one player cooperates, the other does not"
assert score_pdil(p0_coop=False, p1_coop= False, verbose= True) == "both players do not cooperate"


## Part 6 (15 points): Friendly vs Copycat in repeated Prisoner's Dilemma

Complete the `repeated_pdil()` function below which simulates the outcome of a game of repeated prisoner's dilemma between a friendly and copycat player.

In [9]:
def repeated_pdil(score_to_win, coop_friendly_init=True):
    """ runs repeated prisoners dilemma friendly vs copycat
    
    game ends when one player reaches score_to_win.
    
    note that copycat strategy requires the opponents last choice to
    produce a new cooperation decision.  In the first round we use
    coop_friendly_init as the friendly players "previous round" to
    determine coop_copycat, the copycat's current round cooperation
    
    Args:
        score_to_win (int): min threshold to declare (at least)
            one player a winner and end game
        coop_friendly_init (bool): initial condition of friendly
            players "previous round"
            
    Returns:
        score_copycat (int): total copycat player points
        score_friendly (int): total friendly player points
        num_rounds (int): number of rounds played
    """
    #the score of two players are initialized as zero
    score_copycat = 0
    score_friendly = 0
    #they all initialized as True
    copycatpayer = True
    friendlypayer = True

    # track down how many times the game plays
    num_rounds = 0
    

    #in the while loop, while the scores are less than the threshold,
    #keep playing the game
    while score_copycat < score_to_win and score_friendly < score_to_win:
        
        
        #the copycatplayer always follow the previous one
        copycatpayer = get_coop_copycat(coop_friendly_init)

        #as defined in the previous get_coop_friendly funciton 
        #0 represent the false, and 1 represent True
        n = get_coop_friendly()[0]

        if n == 1:
            friendlypayer = True
            coop_friendly_init = True
        else:
            friendlypayer = False
            coop_friendly_init = False


        # simulated the scores of each player
        score_friendly =  score_friendly + score_pdil(friendlypayer,copycatpayer)[0]

        score_copycat = score_copycat + score_pdil(friendlypayer,copycatpayer)[1]

        num_rounds = num_rounds + 1

        

    return score_copycat, score_friendly, num_rounds


In [82]:
repeated_pdil(1000, coop_friendly_init=True)


(1001, 1001, 526)

## Part 7 (12 points): Is one strategy better than another?
To test this, lets run 1000 games of repeated prisoner's dilemman, using the `repeated_pdil()`.  For each game let `score_to_win=20`.  We'll keep track of the outcomes of each game by counting the number of games which have a particular pair of final score tuples `(score_copycat, score_friendly)` in a dictionary called `score_count_dict`.  For example:

```python
score_count_dict = {(20, 0): 999, 
                    (20, 1): 1}
```

indicates that, of the 1000 total games:
- 999 ended with copycat having a score of 20 while friendly has a score of 0
- 1 ended with copycat having a score of 20 while friendly has a score of 1

Note that `score_count_dict` only includes entries for game totals which were observed, none of the values are `0`.

For (+3) extra credit, use a [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict) object for `score_count_dict`.  

In [11]:
from collections import defaultdict

In [23]:
# initialize the variables, the dictionary to store the data
score_count_dict = defaultdict()
score_to_win = 20


# play the game for 1000 times
for i in range(1000):

    result  = repeated_pdil(20,coop_friendly_init=True)

    final_score = (result[0], result[1])
    
    
    # the initial status of the dic
    if final_score not in score_count_dict:
        score_count_dict[final_score] = 1
    # adding up on the dic
    else:
        score_count_dict[final_score] += 1



score_count_dict
     



defaultdict(None,
            {(21, 21): 355,
             (19, 22): 42,
             (20, 20): 496,
             (18, 21): 64,
             (17, 20): 43})

## Part 8 (8 points): Executive Summary
- print the `score_count_dict` you constructed above
- Synthesize the results of your experiment into 1 or 2 clear sentences.  In particular, explain the result to someone who only wants to know:
```
How often does copycat or friendly score more points than the other?  How often do they tie?
```

In [83]:
print(score_count_dict)

defaultdict(None, {(21, 21): 355, (19, 22): 42, (20, 20): 496, (18, 21): 64, (17, 20): 43})


# summary：

as the result of our experiment, we found out that in the turns of 1000 times,
there are 496 + 355 = 851 times that copycat and friendly got the same 
score (tie) while for  149times, friendly got more points and  times .